### PR TITLE
rangefeed: add resolved timestamp initialization scan and registration catch-up scan

### PR DESCRIPTION
**Pull request by nvb**
This PR includes a number of small improvements to the `rangefeed` package, culminating in the introduction of two scans that a rangefeed will need to perform. The first is a scan over the entire key range from the previously known resolved timestamp onward to learn about all existing unresolved intents. The second is a scan over a new registration's key range to backfill in all values from before it was registered but after its starting timestamp. Both of these scans will be able to take advantage of TimeBound iterators. However, the way they're represented in the `rangefeed` package is pretty abstract, which made testing pretty easy.

The first commit and the last two are substantial, but the rest are just small logical changes so I hope this is easy to review.

@benesch thanks for offering to take a look at this while @tschottdorf is out! There's only been one PR merged so far so it shouldn't take much to catch up to speed: https://github.com/cockroachdb/cockroach/pull/28072.

### CODE REVIEW

**Comment by benesch** [comment]
Great, this is in my queue for today!

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/scan.go, line 36 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ04QZ61BdmT2pXK88F:-LJ04QZ61BdmT2pXK88G:b-zgq7qe) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L36)):*
> ```Go
> 	// stop and return the error. If the first result of f is true, the
> 	// iteration stops and returns a nil error.
> 	Iterate(start, end roachpb.Key, f func(engine.MVCCKeyValue) (bool, error)) error
> ```

Why not use the Seek(), Next(), Valid() methods instead, especially given the note in the note in the next commit about needing to add an option to return every version instead of just the latest?
___
*[pkg/storage/rangefeed/scan.go, line 76 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ02j3c6vzqC0Lo9t-Q:-LJ02j3c6vzqC0Lo9t-R:b8uajlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L76)):*
> ```Go
> 	var meta enginepb.MVCCMetadata
> 	sp := s.p.Span.AsRawSpanWithNoLocals()
> 	err := s.snap.Iterate(sp.Key, sp.EndKey, func(kv engine.MVCCKeyValue) (bool, error) {
> ```

the MinTimestampHint is a hint, so I think you can get entries from timestamps outside of the bounds you ask for when you make it. does that matter here? if not, throw a note in here why not, plz

also, nikhil, do you recall what the timebound iterator does with intents? IIRC the original implementation skipped writing the bounds metadata and so that every time bound iterator would have to consider that sst. has that changed?
___
*[pkg/storage/rangefeed/scan_test.go, line 241 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ06Kta2SDb3DbVEVYs:-LJ06Kta2SDb3DbVEVYt:b-360s82) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan_test.go#L241)):*
> ```Go
> 		rangeFeedValue(
> 			roachpb.Key("r"),
> 			roachpb.Value{RawBytes: []byte("val9"), Timestamp: hlc.Timestamp{WallTime: 4}},
> ```

Thoughts on making the lower bound exclusive? If we have a resolved timestamp that we're restarting from, then we've guaranteed that we already emitted the == values, right?

**Review by benesch — COMMENTED** [review]
Wow, #28072 is no joke. Still getting up to speed. I've looked through all the code now and read the RFC and mostly everything looks good. I'll have to do another detailed pass tomorrow though.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 2 of 2 files at r5, 3 of 3 files at r6, 2 of 2 files at r7, 6 of 6 files at r8, 5 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-:-LJ7QNDj48DmafEw6veG:b-zb4j5a)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by benesch — COMMENTED** [review]
Didn't go through the tests with a fine-toothed comb, but got through everything else and have no big concerns. Nice work making this code incredibly readable! It's so refreshing to see small, well-tested objects. :)

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 2 of 2 files at r5, 3 of 3 files at r6, 2 of 2 files at r7, 6 of 6 files at r8, 5 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-:-LJCIws87vGx70U3bXVG:b-gvj5r)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBu9om60ntGRxhcHc0:-LJBu9om60ntGRxhcHc1:b89xf2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85ae7560bed75578480ee468b6a306acecd7353/pkg/storage/rangefeed/processor.go#L296)):*
> ```Go
> // syncEventC synchronizes access to the Processor goroutine, allowing events
> // on its goroutine to establish causality with events on the caller of this
> // method's goroutine.
> ```

I find the comment on this method confusing, and I can't quite put my finger on why. An example would maybe be helpful? 

I'm thinking of this function as something that "flushes" the event pipeline. If you've sent the processor a few events, when `syncEventC` is called, all those events have been processed. Can you check my understanding?
___
*[pkg/storage/rangefeed/processor.go, line 235 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBvnDK0fxprqcuwviR:-LJBvnDK0fxprqcuwviS:b-6g91un) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf912da941fa24d35325d0ae5cad2a6a88b40425/pkg/storage/rangefeed/processor.go#L235)):*
> ```Go
> ) {
> 	// Synchronize the event channel so that this registration doesn't see any
> 	// events that were consumed before it. Instead, it should see these events
> ```

s/before it/before this method was called/
___
*[pkg/storage/rangefeed/processor.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC3di87LGlk16klnV0:-LJC3di87LGlk16klnV1:bqin5u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/processor.go#L126)):*
> ```Go
> 			}
> 		} else {
> 			p.initResolvedTS(ctx)
> ```

This is just for tests, right? If so, I'd prefer if you created a mock snapshot instead. We have these test-only branches all over the place in `storage` and they really add up over time.
___
*[pkg/storage/rangefeed/processor.go, line 175 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC6k9T2r-TqtSQ8in6:-LJC6k9T2r-TqtSQ8in7:b-7esevx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/processor.go#L175)):*
> ```Go
> 					}
> 				} else {
> 					p.handleCatchUpScanRes(ctx, catchUpResult{r: &r})
> ```

Ditto here.
___
*[pkg/storage/rangefeed/processor_test.go, line 402 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBwkAe3HZ7Ej31YUhB:-LJBwkAe3HZ7Ej31YUhC:b-f6dc67) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf912da941fa24d35325d0ae5cad2a6a88b40425/pkg/storage/rangefeed/processor_test.go#L402)):*
> ```Go
> 	go func() {
> 		defer wg.Done()
> 		for i := int64(1); i < 10000; i++ {
> ```

10k ops turns out to be a lot for the race detector. Can you make sure this isn't insanely slow under race?
___
*[pkg/storage/rangefeed/registry.go, line 47 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9AU9EtQ2Y0vTT7Zt:-LJC9AU9EtQ2Y0vTT7Zu:bdudc0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/registry.go#L47)):*
> ```Go
> 	defer s.sendMu.Unlock()
> 	return s.Stream.Send(e)
> }
> ```

Can you get rid of the lock in `testStream` if you're going to lock here?

FWIW it seems reasonable to require that `Send` is threadsafe.
___
*[pkg/storage/rangefeed/registry_test.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJByy3C5Beoa-fSh9jQ:-LJByy3C5Beoa-fSh9jR:bym2tzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/db79b37700cec5aad17983155b6fc4f1e69f945a/pkg/storage/rangefeed/registry_test.go#L100)):*
> ```Go
> 		registration: registration{
> 			span:     span,
> 			caughtUp: true,
> ```

Seems a little funny that test registrations start out caught up, but happy to defer to you here.
___
*[pkg/storage/rangefeed/scan.go, line 76 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ02j3c6vzqC0Lo9t-Q:-LJCIBbxCNvNfQjJuMoZ:bus7rwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L76)):*
> also, nikhil, do you recall what the timebound iterator does with intents? IIRC the original implementation skipped writing the bounds metadata and so that every time bound iterator would have to consider that sst. has that changed?

I remember discussing this, and I thought we settled on requiring that a single intent prevent the bounds metadata from getting written. I cannot find code that does that _anywhere_, though. Unless I'm missing something, I think bounds are currently written as if intents don't exist. :grimacing:  Is backup subtly but completely wrong right now???
___
*[pkg/storage/rangefeed/scan.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9P0G4WX5XAFYMSc5:-LJC9P0G4WX5XAFYMSc6:b-crw675) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan.go#L184)):*
> ```Go
> func (s *catchUpScan) Cancel() {
> 	s.snap.Close()
> }
> ```

I'm somewhat skeptical of these `scan` objects. All the other structs seem spot-on, but these two are super tightly coupled to `Processor`. Do they buy you anything over methods on `Processor` that you would call within a `RunAsyncTask`?
___
*[pkg/storage/rangefeed/scan_test.go, line 241 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ06Kta2SDb3DbVEVYs:-LJC8RSc0r33niR2MHKj:b-7frt5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan_test.go#L241)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Thoughts on making the lower bound exclusive? If we have a resolved timestamp that we're restarting from, then we've guaranteed that we already emitted the == values, right?
</blockquote></details>

To Dan's point: this is what we do in incremental backup (which I think is quite analogous to here). The lower bound is exclusive and the upper bound is inclusive.

**Review by nvb — COMMENTED** [review]
Thanks for the reviews! I made a few changes in the previous commits but tried to keep the majority of large changes to new commits to help with the review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-:-LJG73lS7ghViki-nvV1:by5awd9)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBu9om60ntGRxhcHc0:-LJFOqjg0qWiywobdXyz:b-bq62vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85ae7560bed75578480ee468b6a306acecd7353/pkg/storage/rangefeed/processor.go#L296)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I find the comment on this method confusing, and I can't quite put my finger on why. An example would maybe be helpful? 

I'm thinking of this function as something that "flushes" the event pipeline. If you've sent the processor a few events, when `syncEventC` is called, all those events have been processed. Can you check my understanding?
</blockquote></details>

Yes, that's exactly correct. Want me to change this to `flushEventC` and frame the comment in terms of that?
___
*[pkg/storage/rangefeed/processor.go, line 235 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBvnDK0fxprqcuwviR:-LJFNlrt0Wfn9JPLajph:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf912da941fa24d35325d0ae5cad2a6a88b40425/pkg/storage/rangefeed/processor.go#L235)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

s/before it/before this method was called/
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/processor.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC3di87LGlk16klnV0:-LJFO1P_19knOnuFHR2Y:b-yj7ppw) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/processor.go#L126)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This is just for tests, right? If so, I'd prefer if you created a mock snapshot instead. We have these test-only branches all over the place in `storage` and they really add up over time.
</blockquote></details>

Yeah, I agree with what you're saying and that's what I wanted to do at first. The reason I went with this approach in the end is that a mock snapshot will still be processed asynchronously. This makes testing this reliably harder without even more hooks to enforce causality.

I'm curious if you have any suggestions about how to avoid this extra complication.
___
*[pkg/storage/rangefeed/processor_test.go, line 402 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBwkAe3HZ7Ej31YUhB:-LJFPQe5DvWRBfKA6qVe:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf912da941fa24d35325d0ae5cad2a6a88b40425/pkg/storage/rangefeed/processor_test.go#L402)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

10k ops turns out to be a lot for the race detector. Can you make sure this isn't insanely slow under race?
</blockquote></details>

Good point, done.
___
*[pkg/storage/rangefeed/registry.go, line 47 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9AU9EtQ2Y0vTT7Zt:-LJFQFNJ5ivGWNTeeAQh:b7kp18n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/registry.go#L47)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can you get rid of the lock in `testStream` if you're going to lock here?

FWIW it seems reasonable to require that `Send` is threadsafe.
</blockquote></details>

Yeah, we should just enforce that `Send` is threadsafe. No need to leak gRPC details in here. Done.
___
*[pkg/storage/rangefeed/registry_test.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJByy3C5Beoa-fSh9jQ:-LJFRg3J-mtvrEHPV7xr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db79b37700cec5aad17983155b6fc4f1e69f945a/pkg/storage/rangefeed/registry_test.go#L100)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Seems a little funny that test registrations start out caught up, but happy to defer to you here.
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/scan.go, line 36 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ04QZ61BdmT2pXK88F:-LJG-m_MBqLPqH4Thet7:b-6ln1sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L36)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Why not use the Seek(), Next(), Valid() methods instead, especially given the note in the note in the next commit about needing to add an option to return every version instead of just the latest?
</blockquote></details>

Good point. Switched from this `Snapshot` interface to a standard `engine.SimpleIterator`
___
*[pkg/storage/rangefeed/scan.go, line 76 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ02j3c6vzqC0Lo9t-Q:-LJG-whd2iOyEGJvgSF6:b-5oy60a) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L76)):*
> does that matter here?

No, added a comment about why. It does matter in the catch-up scan though, and we detect that case.

> I think bounds are currently written as if intents don't exist.

Is that a problem for this approach? Does it just make TimeBoundIterators more pessimistic?
___
*[pkg/storage/rangefeed/scan.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9P0G4WX5XAFYMSc5:-LJG0N4sC9QNhL4Wh2hw:b-fwox9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan.go#L184)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I'm somewhat skeptical of these `scan` objects. All the other structs seem spot-on, but these two are super tightly coupled to `Processor`. Do they buy you anything over methods on `Processor` that you would call within a `RunAsyncTask`?
</blockquote></details>

I'm not super attached to them, but I do think they're preferable to methods on Processors. A few reasons why are:
- RunAsyncTask would need to create a closure to capture the iterator or registration, so we're not adding any extra structure, just making it more explicit.
- I think they resulted in a better decomposition of code that's easier to read. That's subjective though so you probably have a better feel for this than I do since I wrote it.
- They're somewhat easier to test and doing so with separate structs requires reaching inside of Processor less.
- They're interacting with the Processor off of its main goroutine through its external API (e.g. calling `ConsumeLogicalOps`), so it seems ok for them to be split off.

Again though, not attached and happy to hear alternatives.
___
*[pkg/storage/rangefeed/scan_test.go, line 241 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ06Kta2SDb3DbVEVYs:-LJG1vFwEuMTOaFyBRI5:b-9w6zuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan_test.go#L241)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

To Dan's point: this is what we do in incremental backup (which I think is quite analogous to here). The lower bound is exclusive and the upper bound is inclusive.
</blockquote></details>

Great point, done.

**Review by danhhz — COMMENTED** [review]
High level LGTM. Deferring on the detailed review to benesch

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-:-LJGHhE40HBmklPRMCMy:babw614)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/scan.go, line 48 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJGH_j_C4p7N_iVAdFp:-LJGH_j_C4p7N_iVAdFq:b7b7cwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e266355e52242c9bcecd9c8896db1e6830f7b0e/pkg/storage/rangefeed/scan.go#L48)):*
> ```Go
> //   An important implication of this is that if the iterator is a
> //   TimeBoundIterator, its MinTimestamp cannot be above the keyspan's largest
> //   known resolved timestamp, if one has ever been recorded. if one has never
> ```

nit: "if one has never" -> "If one has never"

**Review by benesch — APPROVED** [review]
:lgtm:

> Reviewed 8 of 8 files at r10, 3 of 3 files at r11, 2 of 2 files at r12, 6 of 6 files at r13, 6 of 6 files at r14, 1 of 1 files at r15, 6 of 6 files at r16, 4 of 4 files at r17, 5 of 5 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-:-LJHdFQr6rDCRJYf3ZtI:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/rangefeed/processor.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBu9om60ntGRxhcHc0:-LJHXLDdElpEWPDp_wO7:bnt4u1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85ae7560bed75578480ee468b6a306acecd7353/pkg/storage/rangefeed/processor.go#L296)):*
I think my beef is just with the causality talk:

> allowing events on its goroutine to establish causality with events on the caller of this method's goroutine

That description seems backwards to me? Like, isn't it the caller of this method that wants to establish causality? Maybe I'm just misunderstanding?
___
*[pkg/storage/rangefeed/registry_test.go, line 83 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJH_8g0CsS1lOiFO2Vn:-LJH_8g0CsS1lOiFO2Vo:b-6afqhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/795d0ef870b882cdc8a28160fdaca77290dd9ca6/pkg/storage/rangefeed/registry_test.go#L83)):*
> ```Go
> 			panic(err)
> 		}
> 	}
> ```

protoutil.Clone?
___
*[pkg/storage/rangefeed/scan.go, line 76 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ02j3c6vzqC0Lo9t-Q:-LJH_HfSDBAFGSjW1T8Z:b1p6ior) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L76)):*
> Is that a problem for this approach? Does it just make TimeBoundIterators more pessimistic?

No, I think it makes them wrong. You request a scan from [t3, t7] and a time bound iterator might accidentally skip over an SST with keys from t1 and t2 but also an intent at t2. 
___
*[pkg/storage/rangefeed/scan.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9P0G4WX5XAFYMSc5:-LJH_Twq7vpXE-wrU21M:b63ioht) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan.go#L184)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not super attached to them, but I do think they're preferable to methods on Processors. A few reasons why are:
- RunAsyncTask would need to create a closure to capture the iterator or registration, so we're not adding any extra structure, just making it more explicit.
- I think they resulted in a better decomposition of code that's easier to read. That's subjective though so you probably have a better feel for this than I do since I wrote it.
- They're somewhat easier to test and doing so with separate structs requires reaching inside of Processor less.
- They're interacting with the Processor off of its main goroutine through its external API (e.g. calling `ConsumeLogicalOps`), so it seems ok for them to be split off.

Again though, not attached and happy to hear alternatives.
</blockquote></details>

Yeah, I definitely like how self-contained they are, but I think you give up some of that self-containment by having the structs reach back into processor/registry. It seems like you'd want to abstract their communication with processor/registry (maybe through an interface?), but then I definitely think we're in the land of too much abstraction.

Carry on for now though! This is super minor and we can always revisit later.

**Review by benesch — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC3di87LGlk16klnV0:-LJHerRh6JqGJMXQHTV2:b-k3u2yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/processor.go#L126)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I agree with what you're saying and that's what I wanted to do at first. The reason I went with this approach in the end is that a mock snapshot will still be processed asynchronously. This makes testing this reliably harder without even more hooks to enforce causality.

I'm curious if you have any suggestions about how to avoid this extra complication.
</blockquote></details>

Ah, yeah, that is irritating. In the case of the init RTS scan though it seems like syncEventC is already enough of a sync point because of the way the initRTS gets sent over the eventCh?

Regardless, definitely punt on this for now.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28229)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/rangefeed/processor.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJBu9om60ntGRxhcHc0:-LJKRcHI6OFaRYdqEOYu:bmtshk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f85ae7560bed75578480ee468b6a306acecd7353/pkg/storage/rangefeed/processor.go#L296)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think my beef is just with the causality talk:

> allowing events on its goroutine to establish causality with events on the caller of this method's goroutine

That description seems backwards to me? Like, isn't it the caller of this method that wants to establish causality? Maybe I'm just misunderstanding?
</blockquote></details>

Yeah, this was worded poorly. Done.
___
*[pkg/storage/rangefeed/registry_test.go, line 83 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJH_8g0CsS1lOiFO2Vn:-LJKRhru7eiuEn2QJHxQ:b1qcuqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/795d0ef870b882cdc8a28160fdaca77290dd9ca6/pkg/storage/rangefeed/registry_test.go#L83)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

protoutil.Clone?
</blockquote></details>

I wanted to use that, but apparently Go didn't want me to: https://github.com/gogo/protobuf/pull/146 + https://github.com/gogo/protobuf/issues/147.
___
*[pkg/storage/rangefeed/scan.go, line 76 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJ02j3c6vzqC0Lo9t-Q:-LJKSVtk0F0hukxd2Bgf:bhxna4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/26641220807769ed89964e8d7c84052da019ed51/pkg/storage/rangefeed/scan.go#L76)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

> Is that a problem for this approach? Does it just make TimeBoundIterators more pessimistic?

No, I think it makes them wrong. You request a scan from [t3, t7] and a time bound iterator might accidentally skip over an SST with keys from t1 and t2 but also an intent at t2. 
</blockquote></details>

Yikes. I'm going to continue pulling on this thread, but don't want it to hold up this PR. Thanks for bringing it to my attention.
___
*[pkg/storage/rangefeed/scan.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJC9P0G4WX5XAFYMSc5:-LJKTB9c7PdWpHZslMKj:b-1skd4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcb02fbece9553de10fd3a6ebb06f3ef809f3d74/pkg/storage/rangefeed/scan.go#L184)):*
> It seems like you'd want to abstract their communication with processor/registry (maybe through an interface?), but then I definitely think we're in the land of too much abstraction.

Yeah, that's exactly where I landed. I'll wait and see if future development exacerbates the downsides of this decision or not. For now, I'll leave as is.
___
*[pkg/storage/rangefeed/scan.go, line 48 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/28229#-LJGH_j_C4p7N_iVAdFp:-LJKT5QQ1I3TIC9toHt1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e266355e52242c9bcecd9c8896db1e6830f7b0e/pkg/storage/rangefeed/scan.go#L48)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "if one has never" -> "If one has never"
</blockquote></details>

Done.

**Comment by nvanbenschoten** [comment]
Flake https://github.com/cockroachdb/cockroach/issues/26304.

bors r+

**Pull request merged**
